### PR TITLE
shutdown: lazily umount all API vfs

### PR DESCRIPTION
**Pull request by poettering**
Alternative for df48b430a4a85f923eaecb3fadf9c514692d2082.

@medhefgo does this work for you?

### CODE REVIEW

File: src/shutdown/umount.c [hunk 1]
```diff
@@ -158,6 +158,10 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
                         .remount_options = TAKE_PTR(remount_options),
                         .remount_flags = remount_flags,
                         .try_remount_ro = try_remount_ro,
+
+                        /* Unmount sysfs/procfs/… lazily, since syncing doesn't matter there, and it's OK if
+                         * something keeps an fd open to it. */
+                        .umount_lazily = fstype_is_api_vfs(fstype),
```
Comment (medhefgo) [inline src/shutdown/umount.c]: You're calling this twice now in this function now. This should be cached.

File: src/shutdown/umount.c [hunk 2]
```diff
@@ -632,23 +636,14 @@ static int umount_with_timeout(MountPoint *m, bool last_try) {
                  * -EIO rather than blocking indefinitely. If the filesysten is "busy", this may allow
                  * processes to die, thus making the filesystem less busy so the unmount might succeed
                  * (rather than return EBUSY). */
-                r = RET_NERRNO(umount2(m->path, MNT_FORCE));
+                r = RET_NERRNO(umount2(m->path,
+                                       UMOUNT_NOFOLLOW | /* Don't follow symlinks: this should never happen unless our mount list was wrong */
+                                       (m->umount_lazily ? MNT_DETACH : MNT_FORCE)));
```
Comment (medhefgo) [inline src/shutdown/umount.c:659]: Maybe make a `umount_flags` field instead, like `remount_flags`?

**Review by medhefgo — COMMENTED** [review]
I was about to take a shot at this but you were faster. :D

I'll give this a try later.

File: src/shutdown/umount.c (hunk 2 above)
Comment (poettering) [reply src/shutdown/umount.c:659]: I thought about this, but I think we probably should start mangling the flags in the loop here sooner or later, i.e. first try to unmount without MNT_FORCE, and then add this in in a later loop iteration. Thus pre-calculating the flags early doesn't really make too much sense, if you have to mangle them later anyway.

But I didn't want to add the conditional MNT_FORCE stuff into the current PR already, and do that some other time, as it is unrelated.

File: src/shutdown/umount.c [hunk 3]
```diff
@@ -151,18 +151,20 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
                         remount_flags = (remount_flags|MS_REMOUNT|MS_RDONLY) & ~MS_BIND;
                 }
 
-                m = new0(MountPoint, 1);
+                m = new(MountPoint, 1);
                 if (!m)
                         return log_oom();
 
+                *m = (MountPoint) {
+                        .remount_options = TAKE_PTR(remount_options),
```
Comment (medhefgo) [inline src/shutdown/umount.c]: Not like that'll ever matter, but this gets leaked if the strdup futher down fails, no?

Comment (poettering) [reply src/shutdown/umount.c]: nope, the `m` variable is marked with `_cleanup_`

Comment (medhefgo) [reply src/shutdown/umount.c]: Yes, `m` is freed, but `m->remount_options` would leak.

**Comment by medhefgo** [comment]
Seems like this works too.

File: src/shutdown/umount.c (hunk 3 above)
Comment (poettering) [reply src/shutdown/umount.c]: oh, right. it's a simple `_cleanup_free_` indeed

**Pull request merged**